paddle/go/optimizer/optimizer_factory.cc

```diff
@@ -0,0 +1,66 @@
+#include "optimizer_factory.h"
+
+namespace paddle {
+namespace optimizer {
+
+template<class T>
+SGDOptimizer<T>::SGDOptimizer(const ::paddle::OptimizerConfig &config) : ParameterOptimizer<T>(config) {
+  learning_rate = config.learning_rate();
+  momentum = config.momentum();
+  decay = config.decay();
+  nesterov = config.nesterov();
+  lr_decay_a = config.lr_decay_a();
+  lr_decay_b = config.lr_decay_b();
+}
+
+template<class T>
+void SGDOptimizer<T>::destroy() {
+  ~SGDOptimizer();
+}
+
+template<class T>
+void SGDOptimizer<T>::set_weight(const Tensor<T> *p) {
+//  ParameterOptimizer::set_weight(p);
+  size_t size = p->height();
+  // TODO: fix it with align aware allocator bind to Tensor
+  T* ptr = new T[size];
+  momentum_ = Tensor<T>(ptr, size);
+  
+}
+
+template<class T>
+void SGDOptimizer<T>::update(const Tensor<T> &gradient) {
+  num_sample_passed += 1;
+  learning_rate = get_learning_rate();
+  for(size_t i=0; i<parameter_.size(); ++i) {
+    momentums_[i] = momentum * momentums_[i] - learning_rate*gradient[i] - decay*parameter_[i];
```

Comment:
the `momentum`, `nesterov` can be used as switch value. Other framework implements these three optimizers together. These framework has been popular in users. So I just follow their job.
e.g. 
caffe
https://github.com/BVLC/caffe/blob/master/src/caffe/solvers/sgd_solver.cpp#L213
keras(has been the official  frontend use tensorflow and theano)